enumerable/concat.go

```diff
@@ -8,29 +18,49 @@ type enumerableConcat[T any] struct {
 // Concat takes zero to many source `Ènumerable`s and stacks them on top
 // of each other, resulting in one enumerable that will iterate through all
 // the values in all of the given sources.
-func Concat[T any](sources ...Enumerable[T]) Enumerable[T] {
+//
+// New sources may be added after iteration has begun.
+func Concat[T any](sources ...Enumerable[T]) Concatenation[T] {
```

Comment:
suggestion: Not sure if I mentioned this before but we should return concrete type when possible. This way type methods don't need to be strictly tied to the interface.

Reply:
The concrete types are all private, and the new functions public.

Reply:
You can make the concrete types public then. I really think that unless an interface method needs an interface return type that we try to avoid returning interface types

Reply:
> You can make the concrete types public then. I really think that unless an interface method needs an interface return type that we try to avoid returning interface types

I am unsure as to how to express just how much I really do not want to make the concrete types public.

Reply:
Why? You don't have to make the fields public. It's much better than returning an interface type.

Reply:
Nope.  This are internal types, and I highly value the ability to rip and replace them, or allow for more complex constructors which may not return a single concrete type without having to worry about people directly referencing them.

These are not run of the mill business-logic types, but super generic data structures, and I am somewhat sceptical of anyone's ability to convince me to expose them in the short-medium term.  I think that is very bad thing to do.

There is absolutely no reason for anyone to need to directly interact with the concrete types in this package.

Reply:
I like the interfaces (granted I'm a little interface happy compared to average Go dev 😉)

Reply:
> I like the interfaces (granted I'm a little interface happy compared to average Go dev 😉)

The number of times I've hit walls because of us using interfaces for return types 😤... lol

Reply:
For reference: https://github.com/golang/go/wiki/CodeReviewComments#interfaces